### PR TITLE
Enable DMA for NXP LPSPI Driver

### PR DESCRIPTION
**Pull request by danieldegrasse**
This PR enables SPI DMA on all NXP RT10xx platforms, which utilize the LPSPI driver. DMA support should work on other SOCs that have the LPSPI peripheral, but is not currently tested nor enabled in those devicetrees. This PR enables SPI DMA on the following evaluation boards:

- RT1064
- RT1060
- RT1050
- RT1024
- RT1020
- RT1015
- RT1010

### CODE REVIEW

**Review by tbursztyka — APPROVED** [review]

File: tests/drivers/spi/spi_loopback/src/spi.c [hunk 1]
```diff
@@ -35,15 +35,27 @@ struct spi_cs_control spi_cs = {
...
+__aligned(16) uint8_t buffer_rx[BUF_SIZE] __attribute((__section__(".nocache.dma")));
+
+uint8_t BUFFER2_TX[] = "Thequickbrownfoxjumpsoverthelazydog\0";
+__aligned(16) uint8_t buffer2_tx[BUF2_SIZE] __attribute((__section__(".nocache")));
+__aligned(16) uint8_t buffer2_rx[BUF2_SIZE] __attribute((__section__(".nocache.dma")));
+#else
+/* Align buffers in memory to support DMA access */
 uint8_t buffer_tx[] = "0123456789abcdef\0";
-uint8_t buffer_rx[BUF_SIZE] = {};
+uint8_t __aligned(16) buffer_rx[BUF_SIZE] = {};
```
Comment (mmahadevan108) [inline tests/drivers/spi/spi_loopback/src/spi.c:54]: Why are these changes made to the `#else` case?

File: dts/arm/nxp/nxp_rt1010.dtsi [hunk 2]
```diff
@@ -68,6 +68,10 @@
 	interrupts = <38 0>;
 };
 
+&edma0 {
+	dma-channels = <16>;
+};
+
 / {
```
Comment (mmahadevan108) [inline dts/arm/nxp/nxp_rt1010.dtsi:75]: Can this move to a separate commit with details on why this change is needed.

File: boards/arm/mimxrt1064_evk/mimxrt1064_evk.dts [hunk 3]
```diff
@@ -250,8 +250,14 @@ zephyr_udc0: &usb1 {
...
+	/* DMA channels 0 and 1, muxed to LPSPI1 RX and TX */
+	dmas = <&edma0 0 13>, <&edma0 1 14>;
+	dma-names = "rx", "tx";
 };
 
 &lpspi3 {
 	status = "okay";
+	/* DMA channels 2 and 3, muxed to LPSPI3 RX and TX */
+	dmas = <&edma0 2 15>, <&edma0 3 16>;
+	dma-names = "rx", "tx";
```
Comment (mmahadevan108) [inline boards/arm/mimxrt1064_evk/mimxrt1064_evk.dts:262]: Should changes to this file move the commit related to changes to the `boards` folder

File: tests/drivers/spi/spi_loopback/src/spi.c (hunk 1 above)
Comment (danieldegrasse) [reply tests/drivers/spi/spi_loopback/src/spi.c:54]: Those changes were made to ensure that the rx buffer would be aligned for DMA controllers that may require it, as DMA buffers are aligned in the DMA tests in Zephyr. After testing, the EDMA controller in the RT10xx series does not require aligned buffers, so this change has been removed

File: dts/arm/nxp/nxp_rt1010.dtsi (hunk 2 above)
Comment (danieldegrasse) [reply dts/arm/nxp/nxp_rt1010.dtsi:75]: Done, the RT1010 has only 16 channels on its EDMA while other RT10xx platforms have 32.

File: boards/arm/mimxrt1064_evk/mimxrt1064_evk.dts (hunk 3 above)
Comment (danieldegrasse) [reply boards/arm/mimxrt1064_evk/mimxrt1064_evk.dts:262]: Fixed, thank you

File: drivers/spi/spi_mcux_lpspi.c [hunk 4]
```diff
@@ -30,13 +32,38 @@ struct spi_mcux_config {
...
+	struct stream dma_rx;
+	struct stream dma_tx;
+	/* dummy value used for transferring NOP when tx buf is null */
+	uint32_t dummy_tx_buffer;
+	/* dummy value used to read RX data into when rx buf is null */
+	uint32_t dummy_rx_buffer;
+#endif
 };
 
+
```
Comment (mmahadevan108) [inline drivers/spi/spi_mcux_lpspi.c:66]: Minor point, this extra newline can be deleted

File: drivers/spi/spi_mcux_lpspi.c [hunk 5]
```diff
@@ -197,6 +224,232 @@ static int spi_mcux_configure(const struct device *dev,
...
+		      bool asynchronous,
+		      struct k_poll_signal *sig)
+{
+	const struct spi_mcux_config *config = dev->config;
+	struct spi_mcux_data *data = dev->data;
+	LPSPI_Type *base = config->base;
+	int ret;
+	size_t dma_size;
+
+
```
Comment (mmahadevan108) [inline drivers/spi/spi_mcux_lpspi.c:378]: Minor point, one of the new lines can be deleted

**Review by mmahadevan108 — APPROVED** [review]

**Comment by mmahadevan108** [comment]
@danieldegrasse, could you help resolve the CI build errors.

**Comment by cfriedt** [comment]
@danieldegrasse - can you rebase / resolve merge conflicts?

**Comment by danieldegrasse** [comment]
@cfriedt Done, thanks for the heads up

**Comment by danieldegrasse** [comment]
@bbilas Could you take a look at the CI failure? When I bisect it the commit that introduced the failure is 97d3de23. The build failure results from building `tests/drivers/spi/spi_loopback` targeting the `cy8ckit_062_ble_m0` board.

**Comment by bbilas** [comment]
> @bbilas Could you take a look at the CI failure? When I bisect it the commit that introduced the failure is [97d3de2](https://github.com/zephyrproject-rtos/zephyr/commit/97d3de2349500935abbc5e69d140a403cf734208). The build failure results from building `tests/drivers/spi/spi_loopback` targeting the `cy8ckit_062_ble_m0` board.

These 2 MRs should fix those failures:

https://github.com/zephyrproject-rtos/zephyr/pull/40225
https://github.com/zephyrproject-rtos/zephyr/pull/40217

**Comment by cfriedt** [comment]
> > @bbilas Could you take a look at the CI failure? When I bisect it the commit that introduced the failure is [97d3de2](https://github.com/zephyrproject-rtos/zephyr/commit/97d3de2349500935abbc5e69d140a403cf734208). The build failure results from building `tests/drivers/spi/spi_loopback` targeting the `cy8ckit_062_ble_m0` board.
> 
> These 2 MRs should fix those failures:
> 
> #40225 #40226

@bbilas - you could combine them into a single PR. Might be preferable.

**Comment by bbilas** [comment]
> > > @bbilas Could you take a look at the CI failure? When I bisect it the commit that introduced the failure is [97d3de2](https://github.com/zephyrproject-rtos/zephyr/commit/97d3de2349500935abbc5e69d140a403cf734208). The build failure results from building `tests/drivers/spi/spi_loopback` targeting the `cy8ckit_062_ble_m0` board.
> > 
> > 
> > These 2 MRs should fix those failures:
> > #40225 #40226
> 
> @bbilas - you could combine them as separate commits in a single PR. Might be preferable.

I will do that.

**Comment by bbilas** [comment]
@cfriedt @danieldegrasse https://github.com/zephyrproject-rtos/zephyr/pull/40241

**Comment by cfriedt** [comment]
@danieldegrasse - please rebase on a recent `main`. There were some CI changes that went in recently.
